src/codegen/available_expressions.rs

```diff
@@ -0,0 +1,866 @@
+            Rc::new(RefCell::new(BasicExpression {
+                expr_type: expr_type.clone(),
+                expression_id: self.global_id_counter,
+                children: Default::default(),
+            })),
+        );
+
+        expr_set.expr_map.insert(expr_type, self.global_id_counter);
+        self.global_id_counter += 1;
+
+        self.global_id_counter - 1
+    }
+
+    /// Add a node to represent a binary expression
+    pub fn add_expression_node(
+        &mut self,
+        exp: &Expression,
+        expr_set: &mut AvailableExpressionSet,
+        left: NodeId,
+        right: NodeId,
+    ) -> NodeId {
+        let operation = get_operator_from_expression(exp);
+        let new_node = Rc::new(RefCell::new(BasicExpression {
+            expr_type: ExpressionType::BinaryOperation(left, right, operation),
+            expression_id: self.global_id_counter,
+            children: Default::default(),
+        }));
+        expr_set
+            .expression_memory
+            .insert(self.global_id_counter, Rc::clone(&new_node));
+
+        expr_set.expr_map.insert(
+            ExpressionType::BinaryOperation(left, right, operation),
+            self.global_id_counter,
+        );
+
+        expr_set
+            .expression_memory
+            .get_mut(&left)
+            .unwrap()
+            .borrow_mut()
+            .children
+            .insert(self.global_id_counter, Rc::clone(&new_node));
+        expr_set
+            .expression_memory
+            .get_mut(&right)
+            .unwrap()
+            .borrow_mut()
+            .children
+            .insert(self.global_id_counter, Rc::clone(&new_node));
+
+        self.global_id_counter += 1;
+        self.global_id_counter - 1
+    }
+
+    /// Add a node to represent an unary operation
+    pub fn add_unary_node(
+        &mut self,
+        exp: &Expression,
+        parent: usize,
+        expr_set: &mut AvailableExpressionSet,
+    ) -> NodeId {
+        let operation = get_operator_from_expression(exp);
+        let new_node = Rc::new(RefCell::new(BasicExpression {
+            expr_type: ExpressionType::UnaryOperation(parent, operation),
+            expression_id: self.global_id_counter,
+            children: Default::default(),
+        }));
+
+        expr_set
+            .expression_memory
+            .insert(self.global_id_counter, Rc::clone(&new_node));
+
+        expr_set.expr_map.insert(
+            ExpressionType::UnaryOperation(parent, operation),
+            self.global_id_counter,
+        );
+        expr_set
+            .expression_memory
+            .get_mut(&parent)
+            .unwrap()
+            .borrow_mut()
+            .children
+            .insert(self.global_id_counter, Rc::clone(&new_node));
+
+        self.global_id_counter += 1;
+
+        self.global_id_counter - 1
+    }
+}
+
+impl AvailableExpressionSet {
+    /// Check if a commutative expression exists in the set
+    fn check_commutative(
```

Comment:
Rather than return `(bool, NodeId)` I think it would be more rust-like to return `Option<NodeId>`

Reply:
Done